### PR TITLE
devops: deployment pipeline

### PR DESCRIPTION
**Pull request by ternakkode**
The implementation of the AWS infrastructure might seem complex and overkill if it's seen as a solution for just deploying a Layer8 Server. However, some of the implementations here serve as a backbone for any other new product or microservice considering AWS. The implementation offers best practices, secure, cost-efficient, and scalable solutions. When introducing a new service, only a small implementation is required (you'll see no more than 4 new files for a new service to be integrated here).

Deployment Architecture TLDR:

- All AWS resources provisioned via Terraform, Terraform state stored in the Terraform cloud
- Layer8 Server running inside an EC2 Spot Machine, the deployment is managed via ECS (a simplified version of Kubernetes provided by AWS) with 1 GB RAM and 2 Core vCPU
- InfluxDB running inside an EC2 On Demand Machine with 1GB RAM and 2 Core CPU
- DNS and reverse proxy managed by Cloudflare
- InfluxDB and Layer8 Server instances running inside a private subnet without any public IP address. Both instances access the internet via the NAT Instance

Some resources to help with code review : 
- [Terraform Overview](https://www.youtube.com/watch?v=tomUWcQ0P3k&t=35s) to understand basic concepts about terraform and why terraform is a good option.)
- [Detailed video about how terraform work](https://www.youtube.com/watch?v=l5k1ai_GBDE&t=7s) to understand more details about Terraform.
- [Terraform Overview & Syntax Example](https://www.youtube.com/watch?v=HmxkYNv1ksg&t=290s) to understand terraform syntax
- [Article 1](https://www.hashicorp.com/resources/why-consider-terraform-enterprise-over-open-source),  [Article 2](https://medium.com/cloud-native-daily/comparing-terraform-cloud-and-open-source-terraform-for-infrastructure-as-code-27e39d4a5f8d) to understand why terraform cloud is chosen rather than fully using the open source solution
- [ECS Overview from AWS](https://www.youtube.com/watch?v=FnFvpIsBrog) to understand why we should use ECS to manage our layer8 deployment rather than just directly use the EC2 Instance
- [Detailed video about AWS ECS and comparison with Kubernetes](https://www.youtube.com/watch?v=AYAh6YDXuho) to understand more detail about AWS ECS infrastructure
- [Explanation about EC2 Spot 1](https://www.cloudforecast.io/blog/are-aws-spot-instances-worth-it-in-production/)  [Explanation about EC2 Spot 2](https://www.cloudzero.com/blog/spot-instances/) to understand why EC2 SPOT is the best option for Layer8 Server but not for InfluxDB 
- [EC2 Spot Overview video from AWS](https://www.youtube.com/watch?v=aRlY3VBX3Lc) to understand EC2 Spot basic concept
- [Explanation about Cloudflare Reverse Proxy](https://www.cloudflare.com/products/tunnel/) to understand why we use it
- [Basic Concept about how we do networking in AWS](https://www.uturndata.com/2021/02/23/aws-quick-tips-internet-gateways-nat-gateways-and-nat-instances/) to understand the concept of private subnet, public subnet, and NAT
- [Discussion and controversy about why we should have a private subnet](https://stackoverflow.com/questions/22188444/why-do-we-need-private-subnet-in-vpc) to understand why we should put our application in the private subnet besides starting February 2024 AWS costing 2.5$ per month for each public IP address.
- [Discussion started by a user who's annoyed about the high cost of NAT Gateway.](https://www.reddit.com/r/aws/comments/w3zrwz/nat_gateways_are_too_expensive/) to understand why we use NAT Instance rather than NAT Gateway

Cost Estimation per Environment : 
- NAT Instance: 5$
- NAT Public IP Address: 2.5$
- Layer8 Server = 2.30400 (EC2 Spot t3.micro ) + 2.85 (30GB SSD) = 5.1$
- InfluxDB Instance: 7.5 (EC On-Demand t3.micro) + 0.9 (10GB SSD) =  8.4$

Total : 21$ monthly exclude data transfer cost, S3 Storage and ECR storage

### CODE REVIEW

File: Dockerfile
Comment (ternakkode) [inline Dockerfile:30]: When we deploy the image to a service such as ECS/Kubernetes, we no longer use a .env file; instead, we store all configurations in the container environment variables. In this line of code, we add an empty .env file to ensure the application can still run locally, which is easier than putting the configuration in the environment variables.

File: .github/workflows/dev-service.yml
Comment (ternakkode) [inline .github/workflows/dev-service.yml:23]: build docker image and push it to image registry, in this case we use AWS ECR

File: .github/workflows/dev-service.yml
Comment (ternakkode) [inline .github/workflows/dev-service.yml:51]: In this code, we add all configuration or environment variable needed to run the layer8 server to the S3, we are using this to simplify the input of environment variable in the ECS task definition. The ECS will read all value from the S3 then put it to container environment variable automatically

File: .github/workflows/dev-service.yml
Comment (ternakkode) [inline .github/workflows/dev-service.yml:65]: ENV Variable needed as terraform input variable

File: .github/workflows/dev-influxdb.yml
Comment (ternakkode) [inline .github/workflows/dev-influxdb.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/dev-network.yml
Comment (ternakkode) [inline .github/workflows/dev-network.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/prod-influxdb.yml
Comment (ternakkode) [inline .github/workflows/prod-influxdb.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/prod-network.yml
Comment (ternakkode) [inline .github/workflows/prod-network.yml:1]: Basically, it's just a simple script to run Terraform in linux.

File: .github/workflows/prod-service.yml
Comment (ternakkode) [inline .github/workflows/prod-service.yml:1]: same as how we do for dev environment

File: infra/production/base/main.tf
Comment (ternakkode) [inline infra/production/base/main.tf:1]: Define all dependencies needed to run Layer8, InfluxDB, and any new services that may be introduced in the future. This base folder should be moved to a new repository when our other products start using AWS.

In this code, we define:

- Secure and cost-efficient networking solution
- IAM Role to access AWS ECS
- S3 bucket for storing ECS container config
- ECR for storing docker image

File: infra/production/base/module/ecr/main.tf
Comment (ternakkode) [inline infra/production/base/module/ecr/main.tf:8]: Create an ECR Repository for storing Docker Image

File: infra/production/base/module/ecs/capacity_provider.tf
Comment (ternakkode) [inline infra/production/base/module/ecs/capacity_provider.tf:1]: TLDR: This file essentially creates an AWS configuration that automatically generates an AWS EC2 SPOT Instance with specified configurations, alleviating concerns about EC2 Spot Interruptions while still benefiting from cost-effective pricing.

File: infra/production/base/module/ecs/main.tf
Comment (ternakkode) [inline infra/production/base/module/ecs/main.tf:1]: TLDR: This file defines the ECS cluster, IP Whitelist for ECS Instances, and specifies where the ECS can obtain an EC2 instance to run the container.

File: infra/production/base/module/iam/main.tf
Comment (ternakkode) [inline infra/production/base/module/iam/main.tf:1]: TLDR: This file is used to define an AWS IAM Role to grant AWS ECS access to our AWS infrastructure.

File: infra/production/base/module/network/main.tf
Comment (ternakkode) [inline infra/production/base/module/network/main.tf:1]: TLDR: Define VPC with private and public subnet topology and use NAT Instance from open source solution

File: infra/production/influxdb/main.tf
Comment (ternakkode) [inline infra/production/influxdb/main.tf:1]: TLDR: Create an EC2 instance for InfluxDB and execute a script to install InfluxDB and the Cloudflare tunnel inside the EC2 instance.

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:25]: TLDR: task definition is almost similar as docker compose

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:34]: define container for layer8 server

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:50]: define how we store layer8 server logs

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:62]: define cloudflare tunnel as reverse proxy

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:92]: define telegraf as a sidecar to forward metrics from layer8 server to influxdb

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:139]: define the runtime configuration for the layer8 server

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:143]: define how the service horizontally scaled

**Review by ternakkode — COMMENTED** [review]
Comments explaining the implementation

File: infra/development/base/main.tf
Comment (ternakkode) [inline infra/development/base/main.tf:1]: Skip reviewing any file in the infra/development directory, as some of it duplicates code in the infra/production folder. I've added comments to explain the implementation.

File: infra/production/influxdb/main.tf
Comment (ternakkode) [inline infra/production/influxdb/main.tf:30]: You can see this as the ID of the Ubuntu 2022 OS stored in AWS.

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:23]: read data from other terraform workspace

File: infra/production/layer8/main.tf
Comment (ternakkode) [inline infra/production/layer8/main.tf:166]: define the source of ec2 instance to run the container

**Comment by huzaifamk** [comment]
Great expalanation @ternakkode , well done!

**Review by huzaifamk — APPROVED** [review]
I am done reviewing and learning from this, great explanation from @ternakkode.

@NJavokhir if you also review, you will also learn, please add your review approving the PR once you are done, then I will get this merged.

**Comment by ternakkode** [comment]
> I am done reviewing and learning from this, great explanation from @ternakkode.
> 
> @NJavokhir if you also review, you will also learn, please add your review approving the PR once you are done, then I will get this merged.



> I am done reviewing and learning from this, great explanation from @ternakkode.
> 
> @NJavokhir if you also review, you will also learn, please add your review approving the PR once you are done, then I will get this merged.

Let's get this merged first to unblock our progress. Later, Javokhir still can learn about these changes.

**Pull request merged**
